### PR TITLE
feat: original owner & all addresses of paws

### PR DESCRIPTION
**Pull request by d0c0q**

### CODE REVIEW

File: contracts/OtterPAWBondStakeDepository.sol [hunk 1]
```diff
@@ -237,16 +261,110 @@ contract OtterPAWBondStakeDepository is Ownable, ERC721Holder {
...
-     *  @param _paw address
+     *  @notice add discount for a paw
+     *  @param _nft address
      *  @param _discount uint
+     *  @param _endEpoch uint
      */
-    function setPAWDiscount(address _paw, uint256 _discount)
-        external
-        onlyOwner
+    function addDiscountTerms(
```
Comment (otterclamking) [inline contracts/OtterPAWBondStakeDepository.sol:269]:
```suggestion
    function addDiscountTerm(
```

File: contracts/OtterPAWBondStakeDepository.sol [hunk 2]
```diff
@@ -64,7 +66,11 @@ contract OtterPAWBondStakeDepository is Ownable, ERC721Holder {
 
     mapping(address => Bond) public bondInfo; // stores bond information for depositors
     mapping(address => Discount[]) public discountInfo; // stores discount information for depositor
-    mapping(address => uint256) public pawDiscount; // 500 = 5%
+    mapping(address => DiscountTerms) public discountTerms; // stores discount terms for paw.
+    address[] public pawAddresses; // all paws that have discount
+    uint256 public pawCount; // number of paws that have discount
+
+    mapping(address => mapping(uint256 => address)) public pawOwners;
```
Comment (otterclamking) [inline contracts/OtterPAWBondStakeDepository.sol:73]: Looks like you replace paw with nft, so we could replace all paw to nft

```suggestion
    mapping(address => mapping(uint256 => address)) public nftOwners;
```

File: contracts/OtterPAWBondStakeDepository.sol [hunk 3]
```diff
@@ -237,16 +261,110 @@ contract OtterPAWBondStakeDepository is Ownable, ERC721Holder {
...
+     *  @param _tokenID uint256
+     */
+    function discountOfToken(address _nft, uint256 _tokenID)
+        public
+        view
+        returns (uint256)
+    {
+        return
+            _nft != address(0) &&
+                discountTerms[_nft].discount != 0 &&
```
Comment (otterclamking) [inline contracts/OtterPAWBondStakeDepository.sol:332]: why need to check this?

Comment (d0c0q) [reply contracts/OtterPAWBondStakeDepository.sol:332]: addDiscountTerm(address, discount, 0) means getting endEpoch from note

File: contracts/OtterPAWBondStakeDepository.sol (hunk 1 above)
Comment (d0c0q) [reply contracts/OtterPAWBondStakeDepository.sol]: ac7581019fccaa2a52bbbd7c3cc2bd47d4156b13

File: contracts/OtterPAWBondStakeDepository.sol (hunk 2 above)
Comment (d0c0q) [reply contracts/OtterPAWBondStakeDepository.sol]: https://github.com/OtterClam/otter-contracts/pull/6/commits/64fa641d63625e784265156f211f74fa4d938a52

File: scripts/deployPAWBond.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      address: addresses.NFTS.SAFE_HAND_28DAY_NOTE,
+      discount: 50,
+    },
+    {
+      name: 'SAFE_HAND_90DAY_NOTE',
+      address: addresses.NFTS.SAFE_HAND_90DAY_NOTE,
+      discount: 100,
+    },
+    {
+      name: 'SAFE_HAND_180DAY_NOTE',
```
Comment (otterclamking) [inline scripts/deployPAWBond.js:76]: no discount for safe hand note

**Review by otterclamking — CHANGES_REQUESTED** [review]

File: scripts/deployPAWBond.js (hunk 4 above)
Comment (d0c0q) [reply scripts/deployPAWBond.js]: https://github.com/OtterClam/otter-contracts/pull/6/commits/cf18a0fa9f54f54590d247bc2e75736c2304e606

File: contracts/OtterPAWBondStakeDepository.sol [hunk 5]
```diff
@@ -64,7 +66,11 @@ contract OtterPAWBondStakeDepository is Ownable, ERC721Holder {
 
     mapping(address => Bond) public bondInfo; // stores bond information for depositors
     mapping(address => Discount[]) public discountInfo; // stores discount information for depositor
-    mapping(address => uint256) public pawDiscount; // 500 = 5%
+    mapping(address => DiscountTerms) public discountTerms; // stores discount terms for nft.
+    address[] public nftAddresses; // all nfts that have discount
+    uint256 public nftCount; // number of nfts that have discount
+
+    mapping(address => mapping(uint256 => address)) public nftOwners;
```
[the remark this answers is not in the document]
Comment (d0c0q) [reply contracts/OtterPAWBondStakeDepository.sol]: https://github.com/OtterClam/otter-contracts/pull/6/commits/7901075d3b9153e4e67f9c221f9d9c6d1028fb4a

**Pull request merged**